### PR TITLE
[SYCLomatic][ASM] Support migration of PTX instruction "fence" with thread level on sope '.cta', '.gpu' and '.sys'

### PR DESCRIPTION
**Pull request by tomflinda**
Signed-off-by: chenwei.sun <chenwei.sun@intel.com>

### CODE REVIEW

File: clang/lib/DPCT/RulesAsm/AsmMigration.cpp [hunk 1]
```diff
@@ -1219,6 +1219,44 @@ class SYCLGen : public SYCLGenBase {
     return SYCLGenError();
   }
 
+  bool handle_fence(const InlineAsmInstruction *Inst) override {
+    if (Inst->getNumInputOperands() != 0)
+      return SYCLGenError();
+
+    OS() << MapNames::getClNamespace() << "atomic_fence(";
+    if (Inst->hasAttr(InstAttr::sc) && Inst->hasAttr(InstAttr::cta)) {
+
```
Comment (yihwang-nv) [inline clang/lib/DPCT/RulesAsm/AsmMigration.cpp]: Extra blank line.

File: clang/lib/DPCT/RulesAsm/AsmMigration.cpp [hunk 2]
```diff
@@ -1219,6 +1219,44 @@ class SYCLGen : public SYCLGenBase {
...
+           << MapNames::getClNamespace() << "memory_scope::device";
+    } else if (Inst->hasAttr(InstAttr::acq_rel) &&
+               Inst->hasAttr(InstAttr::sys)) {
+      OS() << MapNames::getClNamespace() << "memory_order::acq_rel,"
+           << MapNames::getClNamespace() << "memory_scope::system";
+    } else {
+      SYCLGenError();
+    }
+
+    OS() << ")";
```
Comment (yihwang-nv) [inline clang/lib/DPCT/RulesAsm/AsmMigration.cpp]:
```suggestion
    OS() << ')';
```

**Review by yihwang-nv — APPROVED** [review]
LGTM

**Review by zhiweij1 — APPROVED** [review]

File: clang/lib/DPCT/RulesAsm/AsmMigration.cpp (hunk 1 above)
Comment (tomflinda) [reply clang/lib/DPCT/RulesAsm/AsmMigration.cpp]: Okay

File: clang/lib/DPCT/RulesAsm/AsmMigration.cpp (hunk 2 above)
Comment (tomflinda) [reply clang/lib/DPCT/RulesAsm/AsmMigration.cpp]: Okay

File: clang/lib/DPCT/SrcAPI/APINames_ASM.inc [hunk 3]
```diff
@@ -67,7 +67,8 @@ ENTRY("dp4a", "dp4a", true, NO_FLAG, P1, "Successful")
 ENTRY("elect", "elect", false, NO_FLAG, P1, "Comment")
 ENTRY("ex2", "ex2", true, NO_FLAG, P1, "Successful")
 ENTRY("exit", "exit", false, NO_FLAG, P1, "Comment")
-ENTRY("fence", "fence", false, NO_FLAG, P1, "Comment")
+ENTRY("fence", "true", false, NO_FLAG, P1,
+      "Only support thread fence on sope '.cta', '.gpu' and '.sys'")
```
Comment (zhimingwang36) [inline clang/lib/DPCT/SrcAPI/APINames_ASM.inc]: sope?  typo?

Comment (tomflinda) [reply clang/lib/DPCT/SrcAPI/APINames_ASM.inc]: typo, scope instead.

**Pull request merged**